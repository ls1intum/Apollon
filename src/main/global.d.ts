declare module '*.css';
declare module 'iink-js';
