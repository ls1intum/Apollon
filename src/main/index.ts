export * from './typings';
export * from './apollon-editor';

export type { Patch } from './services/patcher';
