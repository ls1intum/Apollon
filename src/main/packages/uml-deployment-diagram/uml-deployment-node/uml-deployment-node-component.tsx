import React, { FunctionComponent } from 'react';
import { Text } from '../../../components/controls/text/text';
import { UMLDeploymentNode } from './uml-deployment-node';
import { ThemedPath, ThemedRect } from '../../../components/theme/themedComponents';

export const UMLDeploymentNodeComponent: FunctionComponent<Props> = ({ element, children, fillColor }) => (
  <g>
    <g>
      <ThemedPath
        d={`M 0 ${8} l ${8} -${8} H ${element.bounds.width} l -${8} ${8} Z`}
        strokeColor={element.strokeColor}
        fillColor={fillColor || element.fillColor}
      />
      <ThemedPath
        d={`M ${element.bounds.width} 0 V ${element.bounds.height - 8} l -${8} ${8} V ${8} Z`}
        strokeColor={element.strokeColor}
        fillColor={fillColor || element.fillColor}
      />
      <ThemedRect
        x="0"
        y={8}
        width={element.bounds.width - 8}
        height={element.bounds.height - 9}
        strokeColor={element.strokeColor}
        fillColor={fillColor || element.fillColor}
      />
    </g>
    <Text y={30} fill={element.textColor}>
      {element.stereotype && element.displayStereotype && (
        <tspan x="50%" dy={-8} textAnchor="middle" fontSize="85%">
          {`«${element.stereotype}»`}
        </tspan>
      )}
      <tspan
        x="50%"
        dy={element.stereotype && element.displayStereotype ? 18 : 0}
        textDecoration="underline"
        textAnchor="middle"
      >
        {element.name}
      </tspan>
    </Text>
    {children}
  </g>
);

interface Props {
  element: UMLDeploymentNode;
  fillColor?: string;
  children?: React.ReactNode;
}
