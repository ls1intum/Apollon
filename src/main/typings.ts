import { DeepPartial } from 'redux';
import { Styles } from './components/theme/styles';
import { UMLDiagramType } from './packages/diagram-type';
import { UMLElementSelectorType } from './packages/uml-element-selector-type';
import { UMLElementType } from './packages/uml-element-type';
import { UMLRelationshipType } from './packages/uml-relationship-type';
import { ApollonMode, Locale } from './services/editor/editor-types';
import { Direction } from './services/uml-element/uml-element-port';
import { IBoundary } from './utils/geometry/boundary';
import { IPath } from './utils/geometry/path';

export { UMLDiagramType, UMLElementType, UMLRelationshipType, ApollonMode, Locale };
export type { Styles };

export type ApollonOptions = {
  type?: UMLDiagramType;
  mode?: ApollonMode;
  readonly?: boolean;
  enablePopups?: boolean;
  model?: UMLModel;
  theme?: DeepPartial<Styles>;
  locale?: Locale;
  copyPasteToClipboard?: boolean;
  colorEnabled?: boolean;
  scale?: number;
};

export type Selection = {
  elements: string[];
  relationships: string[];
};

export type UMLModel = {
  version: string;
  type: UMLDiagramType;
  size: { width: number; height: number };
  elements: UMLElement[];
  interactive: Selection;
  relationships: UMLRelationship[];
  assessments: Assessment[];
};

export type UMLModelElementType = UMLElementType | UMLRelationshipType | UMLDiagramType;

export type UMLModelElement = {
  id: string;
  name: string;
  type: UMLModelElementType;
  owner: string | null;
  bounds: IBoundary;
  highlight?: string;
  fillColor?: string;
  strokeColor?: string;
  textColor?: string;
  assessmentNote?: string;
  selectedBy?: UMLElementSelectorType[];
};

export type UMLElement = UMLModelElement & {
  type: UMLElementType;
};

export type UMLRelationship = UMLModelElement & {
  type: UMLRelationshipType;
  path: IPath;
  source: {
    element: string;
    direction: Direction;
  };
  target: {
    element: string;
    direction: Direction;
  };
};

export type UMLClassifier = UMLElement & {
  attributes: string[];
  methods: string[];
};

export type UMLDeploymentNode = UMLElement & {
  stereotype: string;
};

export type UMLPetriNetPlace = UMLElement & {
  amountOfTokens: number;
  capacity: number | string;
};

export type UMLReachabilityGraphMarking = UMLElement & {
  isInitialMarking: boolean;
};

export type UMLAssociation = UMLRelationship & {
  source: UMLRelationship['source'] & {
    multiplicity: string;
    role: string;
  };
  target: UMLRelationship['target'] & {
    multiplicity: string;
    role: string;
  };
};

export type UMLCommunicationLink = UMLRelationship & {
  messages: {
    id: string;
    name: string;
    direction: 'source' | 'target';
  }[];
};

export type FeedbackCorrectionStatus = {
  description?: string;
  status: 'CORRECT' | 'INCORRECT' | 'NOT_VALIDATED';
};

export type Assessment = {
  modelElementId: string;
  elementType: UMLElementType | UMLRelationshipType;
  score: number;
  feedback?: string;
  dropInfo?: any;
  label?: string;
  labelColor?: string;
  correctionStatus?: FeedbackCorrectionStatus;
};

export type ExportOptions = {
  margin?: number | { top?: number; right?: number; bottom?: number; left?: number };
  keepOriginalSize?: boolean;
  include?: string[];
  exclude?: string[];
  scale?: number;
};

export type SVG = {
  svg: string;
  clip: {
    x: number;
    y: number;
    width: number;
    height: number;
  };
};
