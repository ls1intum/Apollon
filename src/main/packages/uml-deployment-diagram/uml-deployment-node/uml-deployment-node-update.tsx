import React, { Component, ComponentType } from 'react';
import { connect, ConnectedComponent } from 'react-redux';
import { Button } from '../../../components/controls/button/button';
import { ColorButton } from '../../../components/controls/color-button/color-button';
import { Divider } from '../../../components/controls/divider/divider';
import { TrashIcon } from '../../../components/controls/icon/trash';
import { Textfield } from '../../../components/controls/textfield/textfield';
import { ModelState } from '../../../components/store/model-state';
import { StylePane } from '../../../components/style-pane/style-pane';
import { styled } from '../../../components/theme/styles';
import { UMLElementRepository } from '../../../services/uml-element/uml-element-repository';
import { AsyncDispatch } from '../../../utils/actions/actions';
import { IUMLDeploymentNode, UMLDeploymentNode } from './uml-deployment-node';

const Flex = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;
`;

type State = { colorOpen: boolean };

class DeploymentNodeUpdate extends Component<Props, State> {
  state = { colorOpen: false };

  private toggleColor = () => {
    this.setState((state) => ({
      colorOpen: !state.colorOpen,
    }));
  };

  render() {
    const { element } = this.props;

    return (
      <div>
        <section>
          <Flex>
            <Textfield value={element.name} onChange={this.onRename} autoFocus />
            <ColorButton onClick={this.toggleColor} />
            <Button color="link" tabIndex={-1} onClick={() => this.props.delete(element.id)}>
              <TrashIcon />
            </Button>
          </Flex>
          <StylePane
            open={this.state.colorOpen}
            element={element}
            onColorChange={this.props.update}
            lineColor
            textColor
            fillColor
          />
        </section>
        <section>
          <Divider />
          <Flex>
            <Textfield value={element.stereotype} onChange={this.onUpdate} />
          </Flex>
        </section>
      </div>
    );
  }

  private onRename = (value: string) => {
    const { element, update } = this.props;
    update<IUMLDeploymentNode>(element.id, { name: value });
  };

  private onUpdate = (value: string) => {
    const { element, update } = this.props;
    update<IUMLDeploymentNode>(element.id, { stereotype: value });
  };
}

type OwnProps = {
  element: UMLDeploymentNode;
};

type StateProps = {};

type DispatchProps = {
  update: typeof UMLElementRepository.update;
  delete: AsyncDispatch<typeof UMLElementRepository.delete>;
};

type Props = OwnProps & StateProps & DispatchProps;

const enhance = connect<StateProps, DispatchProps, OwnProps, ModelState>(null, {
  update: UMLElementRepository.update,
  delete: UMLElementRepository.delete,
});

export const UMLDeploymentNodeUpdate: ConnectedComponent<ComponentType<Props>, OwnProps> = enhance(
  DeploymentNodeUpdate,
);
