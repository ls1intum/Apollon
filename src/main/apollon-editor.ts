import 'pepjs';
import { createElement, createRef, RefObject } from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { DeepPartial, Store } from 'redux';
import { ModelState, PartialModelState } from './components/store/model-state';
import { Styles } from './components/theme/styles';
import { UMLElementType } from './packages/uml-element-type';
import { UMLRelationshipType } from './packages/uml-relationship-type';
import { Application } from './scenes/application';
import { Svg } from './scenes/svg';
import { Actions } from './services/actions';
import { ApollonMode, ApollonView, Locale } from './services/editor/editor-types';
import { UMLDiagram } from './services/uml-diagram/uml-diagram';
import { UMLElementRepository } from './services/uml-element/uml-element-repository';
import * as Apollon from './typings';
import { Dispatch } from './utils/actions/actions';
import { UMLDiagramType, UMLModel } from './typings';
import { debounce } from './utils/debounce';
import { ErrorBoundary } from './components/controls/error-boundary/ErrorBoundary';

export class ApollonEditor {
  /**
   * Returns the current model of the Apollon Editor
   */
  get model(): Apollon.UMLModel {
    if (!this.store) {
      console.error('The application state of Apollon could not be retrieved. The editor may already be destroyed.');
      throw new Error('The application state of Apollon could not be retrieved. The editor may already be destroyed.');
    }
    return ModelState.toModel(this.store.getState());
  }

  /**
   * Sets a model as the current model of the Apollon Editor
   * @param model valid Apollon Editor Model
   */
  set model(model: Apollon.UMLModel) {
    if (!this.store) {
      console.error('The application state of Apollon could not be retrieved. The editor may already be destroyed.');
      throw new Error('The application state of Apollon could not be retrieved. The editor may already be destroyed.');
    }
    const state: PartialModelState = {
      ...ModelState.fromModel(model),
      editor: { ...this.store.getState().editor },
    };
    this.recreateEditor(state);
  }

  /**
   * Sets the diagram type of the current Apollon Editor. This changes the selection of elements the user can chose from on the sidebar.
   * @param diagramType the new diagram type
   */
  set type(diagramType: UMLDiagramType) {
    if (!this.store) {
      console.error('The application state of Apollon could not be retrieved. The editor may already be destroyed.');
      throw new Error('The application state of Apollon could not be retrieved. The editor may already be destroyed.');
    }
    const state: PartialModelState = {
      ...this.store.getState(),
      diagram: new UMLDiagram({
        type: diagramType,
      }),
      elements: undefined,
    };
    this.recreateEditor(state);
  }

  /**
   * Sets the current locale of the Apollon Editor.
   * @param locale supported locale
   */
  set locale(locale: Locale) {
    if (!this.store) {
      console.error('The application state of Apollon could not be retrieved. The editor may already be destroyed.');
      throw new Error('The application state of Apollon could not be retrieved. The editor may already be destroyed.');
    }
    const state = this.store.getState();
    this.options.locale = locale;
    this.recreateEditor(state);
  }

  /**
   * renders a model as a svg and returns it. Therefore the svg is temporarily added to the dom and removed after it has been rendered.
   * @param model the apollon model to export as a svg
   * @param options options to change the export behavior (add margin, exclude element ...)
   * @param theme the theme which should be applied on the svg
   */
  static exportModelAsSvg(
    model: Apollon.UMLModel,
    options?: Apollon.ExportOptions,
    theme?: DeepPartial<Styles>,
  ): Apollon.SVG {
    const div = document.createElement('div');
    const element = createElement(Svg, { model, options, styles: theme });
    const svg = render(element, div);
    const { innerHTML } = div;
    unmountComponentAtNode(div);

    return {
      svg: innerHTML,
      clip: svg.state.bounds,
    };
  }

  selection: Apollon.Selection = { elements: [], relationships: [] };
  private currentModelState?: ModelState;
  private assessments: Apollon.Assessment[] = [];
  private application: RefObject<Application> = createRef();
  private selectionSubscribers: ((selection: Apollon.Selection) => void)[] = [];
  private assessmentSubscribers: ((assessments: Apollon.Assessment[]) => void)[] = [];
  private modelSubscribers: ((model: Apollon.UMLModel) => void)[] = [];
  private discreteModelSubscribers: ((model: Apollon.UMLModel) => void)[] = [];
  private errorSubscribers: ((error: Error) => void)[] = [];

  constructor(private container: HTMLElement, private options: Apollon.ApollonOptions) {
    let state: PartialModelState | undefined = options.model ? ModelState.fromModel(options.model) : {};

    state = {
      ...state,
      diagram: new UMLDiagram({
        ...state.diagram,
        type: options.type,
      }),
      editor: {
        ...state.editor,
        view: ApollonView.Modelling,
        mode: options.mode || ApollonMode.Exporting,
        colorEnabled: options.colorEnabled || false,
        scale: options.scale || 1.0,
        readonly: options.readonly || false,
        enablePopups: options.enablePopups === true || options.enablePopups === undefined,
        enableCopyPasteToClipboard: options.copyPasteToClipboard === true,
        features: {
          hoverable: true,
          selectable: true,
          movable: !options.readonly,
          resizable: !options.readonly,
          connectable: !options.readonly,
          updatable: !options.readonly,
          droppable: !options.readonly,
          alternativePortVisualization: false,
        },
      },
    };

    const element = createElement(Application, {
      ref: this.application,
      state,
      styles: options.theme,
      locale: options.locale,
    });
    const errorBoundary = createElement(ErrorBoundary, { onError: this.onErrorOccurred.bind(this) }, element);
    render(errorBoundary, container, this.componentDidMount);
    try {
      this.currentModelState = this.store?.getState();
    } catch (error) {
      this.currentModelState = undefined;
    }
  }

  /**
   * Destroys the Apollon Editor and unmounts it from its container
   */
  destroy() {
    unmountComponentAtNode(this.container);
  }

  /**
   * Selects the by their id identified UMLElements and UMLRelationships
   * @param selection contains ids of the elements and relationships which should be selected
   */
  select(selection: Apollon.Selection) {
    if (!this.store) return;
    const dispatch = this.store.dispatch as Dispatch;
    dispatch(UMLElementRepository.deselect());
    dispatch(UMLElementRepository.select([...selection.elements, ...selection.relationships]));
  }

  /**
   * Register callback which is executed when the selection of elements and relationships changes
   * @param callback function which is called when selection changes
   * @return returns the subscription identifier which can be used to unsubscribe
   */
  subscribeToSelectionChange(callback: (selection: Apollon.Selection) => void): number {
    return this.selectionSubscribers.push(callback) - 1;
  }

  /**
   * Remove selection subscription, so that the corresponding callback is no longer executed when the selection of elements is changed.
   * @param subscriptionId subscription identifier
   */
  unsubscribeFromSelectionChange(subscriptionId: number) {
    this.selectionSubscribers.splice(subscriptionId);
  }

  /**
   * Register callback which is executed when the assessment of elements and relationships are changed
   * @param callback function which is called when assessment changes
   * @return returns the subscription identifier which can be used to unsubscribe
   */
  subscribeToAssessmentChange(callback: (assessments: Apollon.Assessment[]) => void): number {
    return this.assessmentSubscribers.push(callback) - 1;
  }

  /**
   * Remove assessment subscription, so that the corresponding callback is no longer executed when the assessment of elements are changed.
   * @param subscriptionId subscription identifier
   */
  unsubscribeFromAssessmentChange(subscriptionId: number) {
    this.assessmentSubscribers.splice(subscriptionId);
  }

  /**
   * Register callback which is executed when the model changes
   * @param callback function which is called when the model changes
   * @return returns the subscription identifier which can be used to unsubscribe
   */
  subscribeToModelChange(callback: (model: UMLModel) => void): number {
    return this.modelSubscribers.push(callback) - 1;
  }

  /**
   * Remove assessment subscription, so that the corresponding callback is no longer executed when the assessment of elements are changed.
   * @param subscriptionId subscription identifier
   */
  unsubscribeFromModelChange(subscriptionId: number) {
    this.modelSubscribers.splice(subscriptionId);
  }

  /**
   * Register callback which is executed at the end of each user action and ignores the changes during a user action
   * For example: moving of an element is ignored until user releases the element
   * @param callback function which is called when the model changes
   * @return returns the subscription identifier which can be used to unsubscribe
   */
  subscribeToModelDiscreteChange(callback: (model: UMLModel) => void): number {
    return this.discreteModelSubscribers.push(callback) - 1;
  }

  /**
   * Remove model change subscription, so that the corresponding callback is no longer executed when the model is changed.
   * @param subscriptionId subscription identifier
   */
  unsubscribeFromDiscreteModelChange(subscriptionId: number) {
    this.discreteModelSubscribers.splice(subscriptionId);
  }

  /**
   * Register callback which is executed when an error occurs in the editor. Apollon will try to recreate the latest working state when an error occurs, so that it is less visible to user / less interrupting.
   * A registered callback would be called anyway, giving the full error, so that the application which uses Apollon can decide what to do next.
   * @param callback callback function which is called when an error occurs
   * @return returns the subscription identifier which can be used to unsubscribe
   */
  subscribeToApollonErrors(callback: (error: Error) => void): number {
    return this.errorSubscribers.push(callback) - 1;
  }

  /**
   * Removes error subscription, so that the corresponding callback is no longer executed when an error occurs.
   * @param subscriptionId subscription identifier
   */
  unsubscribeToApollonErrors(subscriptionId: number) {
    this.errorSubscribers.splice(subscriptionId);
  }

  /**
   * exports current model as svg
   * @param options options to change the export behavior (add margin, exclude element ...)
   */
  exportAsSVG(options?: Apollon.ExportOptions): Apollon.SVG {
    return ApollonEditor.exportModelAsSvg(this.model, options, this.options.theme);
  }

  private componentDidMount = () => {
    this.container.setAttribute('touch-action', 'none');

    setTimeout(() => {
      if (this.store) {
        this.store.subscribe(this.onDispatch);
      }
    });
  };

  /**
   * Triggered whenever an action is dispatched which potentially lead to a change in the store / state tree
   * Used to notify all the selection and assessment subscribers of Apollon
   */
  private onDispatch = () => {
    if (!this.store) return;
    const { elements, selected, assessments } = this.store.getState();
    const selection: Apollon.Selection = {
      elements: selected.filter((id) => elements[id].type in UMLElementType),
      relationships: selected.filter((id) => elements[id].type in UMLRelationshipType),
    };

    // check if previous selection differs from current selection, if yes -> notify subscribers
    if (JSON.stringify(this.selection) !== JSON.stringify(selection)) {
      this.selectionSubscribers.forEach((subscriber) => subscriber(selection));
      this.selection = selection;
    }

    const umlAssessments = Object.keys(assessments).map<Apollon.Assessment>((id) => ({
      modelElementId: id,
      elementType: elements[id].type as Apollon.UMLElementType | Apollon.UMLRelationshipType,
      score: assessments[id].score,
      feedback: assessments[id].feedback,
      dropInfo: assessments[id].dropInfo,
    }));

    // check if previous assessment differs from current selection, if yes -> notify subscribers
    if (JSON.stringify(this.assessments) !== JSON.stringify(umlAssessments)) {
      this.assessmentSubscribers.forEach((subscriber) => subscriber(umlAssessments));
      this.assessments = umlAssessments;
    }

    // notfiy that action was done
    this.notifyModelSubscribers();
    this.notifyDiscreteModelSubscribers();
  };

  /**
   * Triggered whenever an action is dispatched which potentially lead to a change in the store / state tree
   * Used to notify all the selection and assessment subscribers of Apollon if the action ends with END or DELETE
   */
  private notifyDiscreteModelSubscribers = () => {
    try {
      // if state not available -> do not emit changes
      if (!this.store) return;
      const model = this.model;
      if (
        // At the end of each update operation there is an action that ends with END except DELETE
        // Function is called with every redux action but only notifies subscribers if the action ends with given words
        this.store.getState().lastAction.endsWith('END') ||
        this.store.getState().lastAction.endsWith('DELETE')
      ) {
        const lastModel = ModelState.toModel(this.store.getState());
        this.discreteModelSubscribers.forEach((subscriber) => subscriber(lastModel));
      }
    } catch (error) {
      // if error occured while getting current state for subscribers -> do not emit changes
      // -> no need to emit latest changes
    }
  };

  private notifyModelSubscribers = debounce(() => {
    try {
      // if state not available -> do not emit changes
      if (!this.store) return;
      const model = this.model;
      const lastModel = this.currentModelState ? ModelState.toModel(this.currentModelState) : null;
      if ((!lastModel && model) || (lastModel && JSON.stringify(model) !== JSON.stringify(lastModel))) {
        this.modelSubscribers.forEach((subscriber) => subscriber(model));
        this.currentModelState = this.store.getState();
      } else {
        this.currentModelState = this.store.getState();
      }
    } catch (error) {
      // if error occured while getting current state for subscribers -> do not emit changes
      // -> no need to emit latest changes
    }
  }, 50);

  private recreateEditor(state: PartialModelState) {
    this.destroy();

    const element = createElement(Application, {
      ref: this.application,
      state,
      styles: this.options.theme,
      locale: this.options.locale,
    });
    const errorBoundary = createElement(ErrorBoundary, { onError: this.onErrorOccurred.bind(this) }, element);
    render(errorBoundary, this.container, this.componentDidMount);
  }

  private onErrorOccurred(error: Error) {
    this.errorSubscribers.forEach((subscriber) => subscriber(error));
    this.restoreEditor();
  }

  private restoreEditor() {
    if (this.currentModelState) {
      const state = {
        ...this.currentModelState,
        hovered: [],
        selected: [],
        moving: [],
        resizing: [],
        connecting: [],
        reconnecting: {},
        updating: [],
      };
      this.recreateEditor(state);
    }
  }

  private get store(): Store<ModelState, Actions> | null {
    return (
      this.application.current &&
      this.application.current.store.current &&
      this.application.current.store.current.state.store
    );
  }
}
