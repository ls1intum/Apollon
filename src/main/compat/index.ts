import { UMLModel } from '../typings';
import { isV2, v2ModeltoV3Model, UMLModelCompat as UMLModelCompatV2 } from './v2';

/**
 *
 * Represents all model versions that can be converted to the latest version.
 *
 */
/*
 *
 * HINT for future maintainers:
 *
 * this should always be the union of compatible model versions, i.e.
 * if the model version is moved to V4, while support for V2 is still required,
 * this should look like the following:
 *
 * ```ts
 * export type UMLModelCompat = UMLModelCompatV2 | UMLModelCompatV3;
 * ```
 */
export type UMLModelCompat = UMLModelCompatV2;

/**
 *
 * Converts a model to the latest version.
 *
 * @param {UMLModelCompat} model model to convert
 * @returns {UMLModel} the converted model
 *
 */
export function backwardsCompatibleModel(model: UMLModelCompat): UMLModel {
  if (isV2(model)) {
    return v2ModeltoV3Model(model);
  } else {
    return model;
  }
}
