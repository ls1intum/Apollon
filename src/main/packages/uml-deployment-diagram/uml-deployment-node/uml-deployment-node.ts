import { DeepPartial } from 'redux';
import { DeploymentElementType, DeploymentRelationshipType } from '..';
import { IUMLContainer } from '../../../services/uml-container/uml-container';
import * as Apollon from '../../../typings';
import { assign } from '../../../utils/fx/assign';
import { UMLPackage } from '../../common/uml-package/uml-package';
import { UMLElementType } from '../../uml-element-type';

export interface IUMLDeploymentNode extends IUMLContainer {
  stereotype: string;
  displayStereotype: boolean;
}

export class UMLDeploymentNode extends UMLPackage implements IUMLDeploymentNode {
  static supportedRelationships = [
    DeploymentRelationshipType.DeploymentAssociation,
    DeploymentRelationshipType.DeploymentDependency,
    DeploymentRelationshipType.DeploymentInterfaceProvided,
    DeploymentRelationshipType.DeploymentInterfaceRequired,
  ];
  type: UMLElementType = DeploymentElementType.DeploymentNode;
  stereotype: string = 'node';
  displayStereotype = true;

  constructor(values?: DeepPartial<IUMLDeploymentNode>) {
    super();
    assign<IUMLDeploymentNode>(this, values);
  }

  serialize(): Apollon.UMLDeploymentNode {
    return {
      ...super.serialize(),
      type: this.type as keyof typeof DeploymentElementType,
      stereotype: this.stereotype,
      displayStereotype: this.displayStereotype,
    };
  }

  deserialize<T extends Apollon.UMLModelElement>(values: T, children?: Apollon.UMLModelElement[]): void {
    const assert = (v: Apollon.UMLModelElement): v is Apollon.UMLDeploymentNode =>
      v.type === DeploymentElementType.DeploymentNode;
    if (!assert(values)) {
      return;
    }

    super.deserialize(values, children);
    this.stereotype = values.stereotype;
    this.displayStereotype = values.displayStereotype;
  }
}
