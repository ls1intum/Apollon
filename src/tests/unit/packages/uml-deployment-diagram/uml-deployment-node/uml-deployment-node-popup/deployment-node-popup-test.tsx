import * as React from 'react';
import { fireEvent } from '@testing-library/react';
import { UMLElement } from '../../../../../../main';
import { UMLDeploymentNode } from '../../../../../../main/packages/uml-deployment-diagram/uml-deployment-node/uml-deployment-node';
import { getRealStore } from '../../../../test-utils/test-utils';
import { wrappedRender } from '../../../../test-utils/render';
import { UMLDeploymentNodeUpdate } from '../../../../../../main/packages/uml-deployment-diagram/uml-deployment-node/uml-deployment-node-update';

describe('test deployment node popup', () => {
  let elements: UMLElement[] = [];
  let deploymentNode: UMLDeploymentNode;

  beforeEach(() => {
    // initialize  objects
    deploymentNode = new UMLDeploymentNode({ id: 'source-test-id' });
    elements.push(deploymentNode);
  });

  it('render', () => {
    const store = getRealStore(undefined, elements);

    const { baseElement } = wrappedRender(<UMLDeploymentNodeUpdate element={deploymentNode} />, { store });
    expect(baseElement).toMatchSnapshot();
  });

  it('delete', () => {
    const store = getRealStore(undefined, elements);

    const { getAllByRole } = wrappedRender(<UMLDeploymentNodeUpdate element={deploymentNode} />, { store: store });
    const buttons = getAllByRole('button');
    buttons[0].click();

    expect(store.getState().elements).not.toContain(deploymentNode.id);
  });

  it('rename', () => {
    const store = getRealStore(undefined, elements);

    const { getAllByRole } = wrappedRender(<UMLDeploymentNodeUpdate element={deploymentNode} />, { store: store });
    const textboxes = getAllByRole('textbox');
    const newName = 'UpdatedName';
    fireEvent.change(textboxes[0], { target: { value: newName } });

    expect(store.getState().elements[deploymentNode.id].name).toEqual(newName);
  });

  it('rename stereotype', () => {
    const store = getRealStore(undefined, elements);

    const { getAllByRole } = wrappedRender(<UMLDeploymentNodeUpdate element={deploymentNode} />, {
      store: store,
    });
    const textboxes = getAllByRole('textbox');
    const updatedStereotype = 'UpdatedStereotype';
    fireEvent.change(textboxes[1], { target: { value: updatedStereotype } });

    const updatedElement = store.getState().elements[deploymentNode.id] as UMLDeploymentNode;

    expect(updatedElement.stereotype).toEqual(updatedStereotype);
  });
});
