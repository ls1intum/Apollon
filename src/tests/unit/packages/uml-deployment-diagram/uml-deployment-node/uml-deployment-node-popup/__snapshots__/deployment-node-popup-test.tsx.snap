// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test deployment node popup render 1`] = `
<body>
  <div>
    <div>
      <section>
        <div
          class="sc-bypJrT hnnasP"
        >
          <input
            class="sc-bmzYkS sc-iHGNWf fEVqYL gacFBO"
            maxlength="100"
            value=""
          />
          <button
            class="sc-eeDRCY sc-koXPp fShmjg dCdHsc"
            color="link"
            tabindex="-1"
          >
            <svg
              class="sc-dhKdcB dMUpVV"
              height="16px"
              viewBox="0 0 128 128"
              width="16px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M74.1387 93.5415L92.8052 64.1677H101.049V65.1153L81.9085 93.5415H74.1387ZM74.1387 35.3625H81.9085L101.049 63.7887V64.7363H92.8052L74.1387 35.3625ZM98.7273 93.5415L117.394 64.1677H125.637V65.1153L106.497 93.5415H98.7273ZM98.7273 35.3625H106.497L125.637 63.7887V64.7363H117.394L98.7273 35.3625Z"
                fill="black"
              />
              <path
                clip-rule="evenodd"
                d="M72.068 34.225C72.4831 33.4695 73.2767 33 74.1387 33H81.9085C82.6946 33 83.4292 33.391 83.8682 34.043L103.009 62.4692C103.271 62.8592 103.411 63.3186 103.411 63.7887V65.1153C103.411 65.5854 103.271 66.0448 103.009 66.4348L83.8682 94.861C83.4292 95.513 82.6946 95.904 81.9085 95.904H74.1387C73.2767 95.904 72.4831 95.4345 72.068 94.679C71.653 93.9235 71.6824 93.0019 72.1447 92.2743L89.8254 64.452L72.1447 36.6297C71.6824 35.9021 71.653 34.9805 72.068 34.225ZM93.7418 67.0988L78.4393 91.1789H80.6511L96.8651 67.0988H93.7418ZM96.8651 61.8052H93.7418L78.4393 37.7251H80.6511L96.8651 61.8052ZM96.6567 34.225C97.0717 33.4695 97.8653 33 98.7273 33H106.497C107.283 33 108.018 33.391 108.457 34.043L127.597 62.4692C127.86 62.8592 128 63.3186 128 63.7887V65.1153C128 65.5854 127.86 66.0448 127.597 66.4348L108.457 94.861C108.018 95.513 107.283 95.904 106.497 95.904H98.7273C97.8653 95.904 97.0717 95.4345 96.6567 94.679C96.2416 93.9235 96.271 93.0019 96.7334 92.2743L114.414 64.452L96.7334 36.6297C96.271 35.9021 96.2416 34.9805 96.6567 34.225ZM118.33 67.0988L103.028 91.1789H105.24L121.454 67.0988H118.33ZM121.454 61.8052H118.33L103.028 37.7251H105.24L121.454 61.8052Z"
                fill="black"
                fill-rule="evenodd"
              />
              <path
                d="M29.2727 93.5415H21.5028L2.36253 65.1153V64.1677H10.5588L29.2727 93.5415ZM29.2727 35.3625L10.5588 64.7363H2.36253V63.7887L21.5028 35.3625H29.2727ZM53.5297 93.5415H45.7599L26.6196 65.1153V64.1677H34.8158L53.5297 93.5415ZM53.5297 35.3625L34.8158 64.7363H26.6196V63.7887L45.7599 35.3625H53.5297Z"
                fill="black"
              />
              <path
                clip-rule="evenodd"
                d="M19.5431 34.043C19.9822 33.391 20.7168 33 21.5028 33H29.2727C30.1352 33 30.9291 33.47 31.344 34.2262C31.7588 34.9824 31.7286 35.9045 31.2652 36.632L13.5411 64.452L31.2652 92.2721C31.7286 92.9995 31.7588 93.9216 31.344 94.6778C30.9291 95.434 30.1352 95.904 29.2727 95.904H21.5028C20.7168 95.904 19.9822 95.513 19.5431 94.861L0.402837 66.4348C0.140261 66.0448 0 65.5854 0 65.1153V63.7887C0 63.3186 0.140261 62.8592 0.402837 62.4692L19.5431 34.043ZM6.54628 67.0988L22.7602 91.1789H24.9662L9.62486 67.0988H6.54628ZM9.62486 61.8052H6.54628L22.7602 37.7251H24.9662L9.62486 61.8052ZM43.8002 34.043C44.2392 33.391 44.9738 33 45.7599 33H53.5297C54.3922 33 55.1861 33.47 55.601 34.2262C56.0158 34.9824 55.9856 35.9045 55.5222 36.632L37.7981 64.452L55.5222 92.2721C55.9856 92.9995 56.0158 93.9216 55.601 94.6778C55.1861 95.434 54.3922 95.904 53.5297 95.904H45.7599C44.9738 95.904 44.2392 95.513 43.8002 94.861L24.6599 66.4348C24.3973 66.0448 24.257 65.5854 24.257 65.1153V63.7887C24.257 63.3186 24.3973 62.8592 24.6599 62.4692L43.8002 34.043ZM30.8033 67.0988L47.0173 91.1789H49.2233L33.8819 67.0988H30.8033ZM33.8819 61.8052H30.8033L47.0173 37.7251H49.2233L33.8819 61.8052Z"
                fill="black"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="sc-eeDRCY sc-koXPp fShmjg dCdHsc"
            color="link"
            tabindex="-1"
          >
            <svg
              class="sc-dhKdcB dMUpVV"
              height="16px"
              viewBox="0 0 448 512"
              width="16px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M440 64H336l-33.6-44.8A48 48 0 0 0 264 0h-80a48 48 0 0 0-38.4 19.2L112 64H8a8 8 0 0 0-8 8v16a8 8 0 0 0 8 8h18.9l33.2 372.3a48 48 0 0 0 47.8 43.7h232.2a48 48 0 0 0 47.8-43.7L421.1 96H440a8 8 0 0 0 8-8V72a8 8 0 0 0-8-8zM171.2 38.4A16.1 16.1 0 0 1 184 32h80a16.1 16.1 0 0 1 12.8 6.4L296 64H152zm184.8 427a15.91 15.91 0 0 1-15.9 14.6H107.9A15.91 15.91 0 0 1 92 465.4L59 96h330z"
              />
            </svg>
          </button>
        </div>
      </section>
      <section>
        <hr
          class="sc-tagGq XeZYh"
        />
        <div
          class="sc-bypJrT hnnasP"
        >
          <input
            class="sc-bmzYkS sc-iHGNWf fEVqYL gacFBO"
            maxlength="100"
            value="node"
          />
        </div>
      </section>
    </div>
  </div>
</body>
`;
